### DISCUSSION TITLE
Unable to establish a connection to query-engine-node-api library. It seems there is a problem with your OpenSSL installation!

### DISCUSSION THREAD

**Discussion by kaykhan** [prisma/prisma#14356] [Q&A] [answered]
I have a new laptop with ubuntu 22.04 installed.

I have an npm script which refreshes prisma client. However when i run this i get an error

**package.json**
```
        "prisma-refresh": "npx prisma db pull && npx prisma generate && npm install @prisma/client"
```

**npm run prisma-refresh**

> 
> Error: Get Config: Unable to establish a connection to query-engine-node-api library. It seems there is a problem with your OpenSSL installation!
> Details: Unable to require(`/home/kay/checkpoint/nft-api/node_modules/@prisma/engines/libquery_engine-debian-openssl-1.1.x.so.node`)
>  libssl.so.1.1: cannot open shared object file: No such file or directory

**prisma/schema.prisma**
```
generator client {
  provider = "prisma-client-js"
}
...
```

 **whereis node**
```
node: /home/kay/.nvm/versions/node/v16.16.0/bin/node
```

**whereis openssl**

```
openssl: /usr/bin/openssl /home/linuxbrew/.linuxbrew/bin/openssl /usr/share/man/man1/openssl.1ssl.gz
```

**openssl version**
```
OpenSSL 1.1.1q  5 Jul 2022
```

**prisma**: 4.0.0

What is the fix for this?

**ATTEMPT 1:**

**echo $PATH**

```
/home/kay/.gvm/bin:/home/kay/.pyenv/shims:/home/linuxbrew/.linuxbrew/Cellar/pyenv-virtualenv/1.1.5/shims:/home/kay/.pyenv/bin:/home/kay/.nvm/versions/node/v16.16.0/bin:/home/kay/bin:/usr/local/bin:/home/linuxbrew/.linuxbrew/bin:/home/linuxbrew/.linuxbrew/sbin:/home/kay/.gvm/bin:/home/linuxbrew/.linuxbrew/Cellar/pyenv-virtualenv/1.1.5/shims:/home/kay/.pyenv/bin:/home/kay/.nvm/versions/node/v16.16.0/bin:/home/kay/bin:/usr/local/bin:/home/kay/.cargo/bin:/home/linuxbrew/.linuxbrew/bin:/home/linuxbrew/.linuxbrew/sbin:/home/kay/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/snap/bin
```

if i set `/usr/bin` to be the first item in the path, openssl will then become version 3.0 and it will work. But i don't know if this is the correct solution? And ive only managed to set this temporarily. THIS IS NOT A GOOD SOLUTION, because it then breaks a lot of programs i depend on that are installed via brew.

**Comment by pantharshit00** [comment]
Hello @kaykhan, our apologies for the delay in getting back to you! Due to the large number of questions in our community channels, unfortunately this slipped through the cracks. It's very important for us to address your question, so let us know in a comment if this is still an open discussion. 

If this is still relevant, can you please make sure you haven't installed Node.js via snap. Snap install sandboxes the openssl install so Prisma Node API library won't be able to find it.

**Reply by vladi-stevanovic** [reply]
👋 Hello @kaykhan I'm going through pending GitHub Discussions to check if you need any help. Is this issue still open?
Please let us know if we can investigate it further or if you've already solved it, we'd love to know how!

It would be super useful for other users facing this problem! 🙌

**Comment by jobyh** [comment, marked as answer]
@vladi-stevanovic for anyone else who finds this discussion I was encountering the same issue when running `npx prism db pull`

I'm running Ubuntu 22.04 with Node 18.12 LTS installed via NVM. I noticed that my OpenSSL version was also 1.1.1q installed by Linuxbrew. 

Unlinking the Linuxbrew version so that the system version of OpenSSL (3.0.2) is used resolved the issue for me and I have been able to run the pull successfully.

```
brew unlink openssl@1.1
```

In my case I had installed 1.1 trying to get a local MongoDB installation working on Ubuntu 22.04 unsuccessfully. Ended up using Atlas.

**Reply by janpio** [reply]
Can you describe a bit further how we could get our system into the same situation? We would like to be able to reproduce this and then hopefully find a better solution.

**Reply by John-Dennehy** [reply]
@jobyh thank you so much - been stuck with the same issue for weeks. I have no idea why I would have a duplicate (and outdated) version installed via homebrew, but gave it a last ditch attempt any way and sure enough it worked!

**Reply by jobyh** [reply]
@John-Dennehy You’re welcome, I’m really pleased it has helped you too! Hoping it doesn’t break any other Homebrew packages which may have installed 1.1 as a dependency. Still, as you say, better not to be using the outdated version.

> Can you describe a bit further how we could get our system into the same situation? We would like to be able to reproduce this and then hopefully find a better solution.

@janpio Yes, of course. In the context of Ubuntu 22.04 (LTS):
* Ensure OpenSSL toolkit is installed `sudo apt-get install openssl`
* Check system OpenSSL is v3 `openssl version`
* Install [Homebrew](https://brew.sh/)
* Install NVM via Homebrew `brew install nvm`
* Install Node LTS via NVM `nvm install —lts`
* Install legacy OpenSSL via Homebrew `brew install openssl@1.1`
* Check legacy OpenSSL is now being used `openssl version`

The above should recreate the system state into which I created a Next.js app and added Prisma which errored on `npx prisma db pull`. If you need detailed steps am happy to provide them.

I don’t think Prisma is doing anything _really_ wrong because the cause is an outdated version of OpenSSL overriding the system version for that user. The problem is very frustrating from a DX perspective as it's not clear what action to take based on the error message so I appreciate you wanting to look for an improvement 🙂

**Reply by Jolg42** [reply]
I did a reproduction on GitHub Actions Ubuntu 22.04 here https://github.com/prisma/ecosystem-tests/pull/3204
Though the error is different
```
[...]
Query engine library for current platform "debian-openssl-3.0.x" could not be found.
You incorrectly pinned it to debian-openssl-3.0.x
[...]
```

**Reply by jobyh** [reply]
@Jolg42 I wonder if the difference in error message could be related to the fact that Homebrew overrides for a specific user on the system (not all users)? Is the action running the command  which triggers the error as a different user than the one whos path includes Homebrew?

**Reply by pantharshit00** [reply]
Hello @jobyh 

I have converted this discussion into an issue so this can be properly tracked. You can follow up here now: https://github.com/prisma/prisma/issues/16420

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
